doc/architecture/environment/deployment-environment.md

# Deployment Environment

We use the [AWS CDK](https://github.com/aws/aws-cdk) to build and deploy the
various parts of the stack.

https://github.com/au-research/raido-v2-aws-private/

The setup and configuration of your deployment process depends heavily on your
[operational environment](./operational-environment.md).

For example, our CloudFront is configured to serve the app-client from S3.
We have a AWS CodeBuild project that can be invoked to run the app-client
build process and deploy the built files to S3.  CloudFront serves those files
over the internet.

How you deploy those files will depend on your environment,
you might:
* configure your web-server to read files from a network drive (via NFS protocol
  or similar)
  * then your build process just needs to copy the files to the expected network
    location

* configure your web-server to read files from its local drive
  * then your build process would need to use `scp` or a similar protocol to 
  copy the files to each web server machine


# ARDC RAiD Service

## AWS infrastructure 

Our AWS infrastructure is deployed via an 
[AWS CDK pipeline](https://docs.aws.amazon.com/cdk/v2/guide/cdk_pipeline.html) 
in a "continuous integration" style  - i.e. AWS infrastructure is automatically 
updated when a change is pushed to the `/main` branch of the AWS repo.


## Database schema

The database instance itself is managed via CDK.
The database schema (tables, columns, triggers, etc.) is implemented as 
[Flyway](https://flywaydb.org/) migrations that are deployed by an 
[AWS CodeBuild](https://aws.amazon.com/codebuild/) task.

* [RaidoDbFlywayMigrate project](https://github.com/au-research/raido-v2-aws-private/blob/fd26c55ab476533e6c3d9c2cd6f712046b101ba1/raido-root/lib/prod/raido/RaidoDbCodeBuild.ts#L69)


## api-svc docker container 

The api-svc container instances are managed as [AWS ECS](https://aws.amazon.com/ecs/) 
tasks.

The docker container that contains the actual api-svc task is built by an 
AWS CodeBuild project: 
[ApiSvcPushDockerImage project](https://github.com/au-research/raido-v2-aws-private/blob/fd26c55ab476533e6c3d9c2cd6f712046b101ba1/raido-root/lib/prod/api-svc/ApisSvcCodeBuild.ts#LL50C53-L50C68)

The core Gradle task that initiates the Java compile and docker build (called 
by the PushDocker image project) is 
[ecsDockerPush](/api-svc/docker/build.gradle)


## app-client build

The app-client is built as a [create-react-app](https://create-react-app.dev/) 
project ([Webpack](https://webpack.js.org/) under the covers),
then deployed via a CodeBuild project:
[AppClientDeploy](https://github.com/au-research/raido-v2-aws-private/blob/fd26c55ab476533e6c3d9c2cd6f712046b101ba1/raido-root/lib/prod/app-client/AppClientCodeBuild.ts#L36)

The core build task is a simple `npm run build`, then the project copies
the built files to S3.




doc/architecture/environment/operational-environment.md

# Operational Environment 

To run the Raido codebase in a non-AWS environment, you would need to consider 
how to provide equivalent services for each of the sections outlined below. 

The below sections were mapped directly from the  
[raid-architecture.md](/doc/architecture/raid-architecture.md).


# ARDC RAiD Service environment in AWS

## DNS
* [AWS Route53](https://aws.amazon.com/route53/)
* https://github.com/au-research/raido-v2-aws-private/blob/main/raido-root/lib/prod/Route53.ts


## TLS certificates
* [AWS Certificate Manager](https://aws.amazon.com/certificate-manager/)
* https://github.com/au-research/raido-v2-aws-private/blob/main/raido-root/lib/prod/Us1Certificate.ts


## External sign-in services

Raido uses [Single sign-on](https://en.wikipedia.org/wiki/Single_sign-on).
Raido is designed to use identity services OAuth2 / OIDC to allow users to 
sign-in with pre-existing credentials.  
Note that Raido currently implements no direct sign-in method 
(i.e. storing username/email/password data in our DB).

Currently, we authenticate uses with the following providers:
* [Google](https://developers.google.com/identity/openid-connect/openid-connect)
* [AAF](https://support.aaf.edu.au/support/solutions/articles/19000096640-openid-connect-)
* [ORCID](https://info.orcid.org/documentation/features/public-api/orcid-as-a-sign-in-option-to-your-system/)

Each of these authentication methods requires an agreement between the Relying 
Party (i.e. ARDC) and the Identity Provider.

See [/doc/security/access-control/authentication](/doc/security/access-control/authentication)
for more detail about how sign-in and acess-control works.


## app-client - front-end web server

The app-client is a React "Single Page App", built as a purely static set of 
files to be loaded by the web browser.  It doesn't need much from the web
server beyond just serving the static files and taking care of TLS.

* [AWS CloudFront](https://docs.aws.amazon.com/cloudfront/index.html)
* https://github.com/au-research/raido-v2-aws-private/blob/main/raido-root/lib/prod/app-client/AppClientCloudFront.ts
* [architecture decision log](/app-client/doc/adr)


###  Load balancing

Front-end Load balancing is completely encapsulated by the CloudFront service.
CloudFront handles caching of files local to the end-user (if configured) and 
handles all routing of user traffic.

We also have some rules set up so that the CloudFront instances forward API 
traffic to the api-svc load balancer.  This is not necessary (and incurs cost
at the Cloudfront/networking level), but it removes
the need for the browser to send pre-flight CORS requests.

See https://github.com/au-research/raido-v2-aws-private/blob/fd26c55ab476533e6c3d9c2cd6f712046b101ba1/raido-root/lib/prod/app-client/AppClientCloudFront.ts#L53


## api-svc - back-end API service

Raido is a Java/Spring API server.
It is completely stateless, designed for horizontal scalability.
* no 2nd level caching or similar
* no use of HTTP session state
* load balancing can be completely "un-sticky"

* [AWS ECS](https://aws.amazon.com/ecs/) running docker containers stored in 
  [AWS ECR](https://aws.amazon.com/ecr/)
* https://github.com/au-research/raido-v2-aws-private/blob/main/raido-root/lib/prod/api-svc/ApiSvcEcs.ts
* [Dockerfile](/api-svc/docker/src/main/docker/Ecs.Dockerfile)
* [architecture decision log](/api-svc/doc/adr/readme.md)


### Load balancing

We use [AWS ALB](https://aws.amazon.com/elasticloadbalancing/application-load-balancer/) 
in combination with ECS for routing API traffic to api-svc containers running 
in ECS.

* https://github.com/au-research/raido-v2-aws-private/blob/main/raido-root/lib/prod/raido/ProdRaidOrgAuAlb.ts
* https://github.com/au-research/raido-v2-aws-private/blob/fd26c55ab476533e6c3d9c2cd6f712046b101ba1/raido-root/lib/prod/api-svc/ApiSvcEcs.ts#L244


### External PID services

Currently, the api-svc validates PIDs (ORCID, ROR, DOI) with external services, 
but it does so using only publicly available HTTP endpoints that do not require
authentication.

There are currently no API keys or other accounts/credentials that need to be
set up to allow this.  

Though the api-svc does requires network access to the internet so that 
http/https requests can be made.


## Local Handle Service

We currently use the ARDC APIDS service for minting handles.
It is operated in an "on-premises" fashion, on ARDC hardware and network 
services, by the ARDC DevOps team (separate from the Raid team).

* https://ardc.edu.au/services/ardc-identifier-services/ardc-handle-service/
* https://github.com/au-research/ANDS-PIDS-Service
* [ApidsService.java](/api-svc/spring/src/main/java/raido/apisvc/service/apids/ApidsService.java)
* There's no AWS CDK code for this since it's operated externally to AWS
  * the configuration is stored in the ECS task definition
  * https://github.com/au-research/raido-v2-aws-private/blob/fd26c55ab476533e6c3d9c2cd6f712046b101ba1/raido-root/lib/prod/api-svc/ApiSvcEcs.ts#LL179C31-L179C31


## Database

Postgres 15 database server operated by [AWS RDS](https://aws.amazon.com/rds/).

* https://github.com/au-research/raido-v2-aws-private/blob/main/raido-root/lib/prod/raido/RaidoDb.ts
* [architecture decision log](/api-svc/doc/adr/2022-07-21-database.md)
* [postgres-specific-features.md](/api-svc/doc/postgres-specific-features.md)


## Log management and observability

The ECS task definition is configured to send logs to 
[AWS Cloudwatch](https://aws.amazon.com/cloudwatch/).

* https://github.com/au-research/raido-v2-aws-private/blob/fd26c55ab476533e6c3d9c2cd6f712046b101ba1/raido-root/lib/prod/api-svc/ApiSvcEcs.ts#L127

### Metrics
* [AWS CloudWatch metrics](https://docs.aws.amazon.com/AmazonCloudWatch/latest/monitoring/working_with_metrics.html)
* [observability.md](/doc/observability.md)


## Secrets management

* [AWS Secrets Manager](https://docs.aws.amazon.com/secretsmanager/index.html)
* https://github.com/au-research/raido-v2-aws-private/blob/main/raido-root/lib/prod/api-svc/ApiSvcSecret.ts
* https://github.com/au-research/raido-v2-aws-private/blob/main/raido-root/lib/prod/Secrets.ts



changelog.md
See the [Changelog audience](#changelog-audience) section for info about 
 the expected audience and content of the changelog.

# Next version

_Not yet scheduled_

## App-client UI
* none
 
## API
* updates to experimental API to align terminology
  * unapproved-user-experimental.yaml renamed from "unauthz"

## Infrastructure
* none


---

# 1.2

_2023-05-15_

## App-client UI
* Additions to mint/edit page:
  * Spatial coverage field
  * Traditional knowledge label field
* Option to disabled editing by service point

## API 
* PID validation
  * check ORCID exists for contributor
  * check ROR exists for organisation
  * check DOI exists for service point
  * note that the service-level-guide has been updated to allow that API 
    requests maybe rejected if they have too many PIDs associated
* New metadata blocks:
  * spatial coverages
  * traditional knowledge labels

## Infrastructure
* introduced the InMemory stubs for APIDS, ORCID, ROR and DOI
  * this is noted in changelog not because it changes anything, but the 
    deployment plan should take it into account, specifically: 
    * attention should be paid to env config (shouldn't need changing)
    * post-deploy tests should verify that we didn't accidentally start using 
      in-memory handles in production
* DB changes 
  * bumped width of `raid.handle` column
  * edit service point new column
* Metrics publishing to AWS CloudWatch introduced
  * must be enabled in config
  * remember to add the permissions to the ECS task role (not the ASG 
    cluster role), see DEMO `ApiSvcEcs`

---

# 1.1

_2023-04-11_

## App-client UI changes 

* Change public landing page title from "Raido" to "RAiD"
* Add sign-in warning of uBlock Origin / ORCID issue
* Add "Copy handle" button to the home page list
* Add download for "recently minted raids" report
* Default the lead organisation on the mint page to the institution
  associated with the service-point.
* Add item to menu for admin users to access API keys

## API changes

New metadata blocks:
* subjects 
* related Raids
* related objects 
* alternate identifiers


# 1.0

_2023-03-15_

Initial production deployment.

---

# Changelog audience

The changelog has multiple audiences:
* communications team - for preparing the notification email that
  will be sent for a new deployment
  * mostly interested in the UI section
  * but they also need to know about deprecations of stable API endpoints
* client API integrators
  * mostly interested in the API section
* test team
  * interested in both UI And API sections
* deployment team
  * interested in the infrastructure section


app-client/src/Config.ts

/** Helps manage configuration values across multiple environments.
 * <p>
 * The config values used by the app are decided when this file is run (i.e.
 * when the app is loaded into the browser), based on the value of
 * environmentName, which is compiled in at build time via the
 * REACT_APP_RAIDO_ENV environment variable.
 * <p>
 * Config values are embedded into the raw client javascript artifact - it is
 * designed only for static, publicly visible config (i.e. not per user stuff
 * and definitely not secrets).
 * <p>
 * Config is not suitable for secrets like API keys or other sensitive data.
 * <p>
 * EnvironmentConfig is things we expect to change between environments.
 * <p>
 * Note, often there'll be lots of config that ends up using the same values
 * across most environments. Avoid the temptation to re-invent the concept of  
 * "shared" environment configuration.  
 * Each config (`prod` / `demo` etc.) should be reviewable as a stand-alone 
 * unit (in a non-IDE context).
 * Don't force reviewers to guess what the `prod` or `test` config looks like 
 * by mentally reproducing your "shared config merging logic".  Preventing 
 * accidental/unexpected usage of bad config (e.g. a test or prod environment 
 * accidentally using a wrong DB or other system dependency) is more important
 * than your personal obsession with the DRY principle.
 */

const log = console;

/** Defines what the known environments are. */
type EnvironmentName = "prod" | "demo" | "ci" | "dev";

export interface AuthnConfig {
  clientId: string,
  authorizeUrl: string,
  authnScope: string,
}

/**
 couldn't figure out how to use QueryObserverOptions for this, so just settled
 on a simplified version cut down for what we actually want to set.
 */
export interface ReactQueryConfig {
  retry: boolean | number,
  refetchOnWindowFocus: boolean | 'always',
}


export interface EnvironmentConfig {
  /** identifies the environment */
  environmentName: EnvironmentName,

  /** is this a "production" environment (usually there's only one)? 
   * If you feel the need to have switchable business logic based off of
   * your environment, this is what it should be predicated on 
   * (i.e not "if config.envName === "prd").
   */
  isProd: boolean,

  signInWarning?: string, 
  aaf: AuthnConfig,
  google: AuthnConfig,
  orcid: AuthnConfig,
  
  raidoIssuer: string,
  raidoLandingPage: string,
  /** The hostname to use to call the api-svc endpoints. 
  Generally empty string, so that API calls are a relative to current host.
  But you can use a different host if desired (but remember to configure CORS
  headers if you want to do that). */
  raidoApiSvc: string,

  authApiQuery?: ReactQueryConfig,
  publicApiQuery?: ReactQueryConfig,
}

function initConfig(){
  const newConfig = {
    ...buildConfig,
    ...chooseEnvironmentConfig(process.env.REACT_APP_RAIDO_ENV),
  };

  log.debug("Application config", process.env.REACT_APP_RAIDO_ENV, newConfig);
  return newConfig;
}

export const unknownCommitId = "unknown commit";
export const unknownBuildDate = "unknown";

const buildConfig = {
  /* expected to be populated from something like 
  `date --iso-8601=seconds --utc`, e.g. `2023-02-01T06:31:13+00:00` */
  buildDate: process.env.REACT_APP_BUILD_DATE_MS ?? "unknownBuildDate" ,
  version: process.env.REACT_APP_RAIDO_VERSION ?? "unknown version",
  gitCommit: process.env.REACT_APP_COMMIT_REF ?? unknownCommitId,
};


function chooseEnvironmentConfig(env: string | undefined){
  /* trim() because env var can get polluted with whitespace when set via
  npm scripts. */
  env = env?.toLowerCase()?.trim();
  if( env === 'prod' ){
    return prodConfig
  }
  else if( env === 'demo' ){
    return demoConfig;
  }
  else if( env === 'ci' ){
    return ciConfig;
  }
  else if( env === 'dev' ){
    return devConfig;
  }
  else {
    console.log("unknown env, using dev: ", env);
    return devConfig;
  }
}

const ciConfig: EnvironmentConfig = {
  environmentName: "ci",
  isProd: false,
  raidoIssuer: "https://demo.raid.org.au",
  raidoLandingPage: "https://demo.raid.org.au",
  raidoApiSvc: "",
  aaf: {
    clientId: "",
    authorizeUrl: "",
    authnScope: "",
  },
  google: {
    clientId: "",
    authorizeUrl: "",
    authnScope: "",
  },
  orcid: {
    clientId: "",
    authorizeUrl: "",
    authnScope: "",
  },
};

const devConfig: EnvironmentConfig = {
  environmentName: "dev",
  isProd: false,
  // have to run the https proxy for this, see local-orcid-signin.md  
  //raidoIssuer: "https://localhost:6080",
  raidoIssuer: "http://localhost:8080",
  raidoLandingPage: "http://localhost:8080",
  raidoApiSvc: "",
  aaf: {
    // https://aaf.freshdesk.com/support/solutions/articles/19000096640-openid-connect-
    clientId: "accaabfd-a7c8-4d36-9363-ea7342e24db5",
    authorizeUrl: "https://central.test.aaf.edu.au/providers/op/authorize",
    authnScope: "openid email profile",
  },
  google: {
    // DEMO OAuth client on raid.services@ardc account
    clientId: "333652356987-ecetlds8nprpf81er55jnva1750jputr.apps.googleusercontent.com",
    authorizeUrl: "https://accounts.google.com/o/oauth2/v2/auth",
    authnScope: "openid email profile",
  },
  // this is not going to work, orcid require https
  orcid: {
    clientId: "APP-IZBIZ6O7XH9RFG0X",
    authorizeUrl: "https://orcid.org/oauth/authorize",
    authnScope: "openid",
  },
  // disable because kind of annoying in a development context
  authApiQuery: {
    retry: false,
    refetchOnWindowFocus: false,
  },
  publicApiQuery: {
    retry: false,
    refetchOnWindowFocus: false,
  },
};

const demoConfig: EnvironmentConfig = {
  environmentName: "demo",
  signInWarning: `The DEMO environment has been reset as of 2023-04-03.
    You may need to submit a new authorisation request to use the system.
    The sign-in expiry time has been set to 30 minutes for testing.
    Please send an email to contact@raid.org to let us know you need approval.`,
  isProd: false,
  raidoIssuer: "https://api.demo.raid.org.au",
  raidoLandingPage: "https://demo.raid.org.au",
  raidoApiSvc: "",
  aaf: {
    clientId: "205290b0-8bd6-4d53-8605-d0237636b235",
    authorizeUrl: "https://central.aaf.edu.au/providers/op/authorize",
    authnScope: "openid email profile",
  },
  google: {
    // OAuth client: raid.service@ardc.edu.au / ardc.edu.au / Raido production / Raido PROD
    clientId: "333652356987-6n1dlouree79qbffrtgqd9fit6phkvs5.apps.googleusercontent.com",
    authorizeUrl: "https://accounts.google.com/o/oauth2/v2/auth",
    authnScope: "openid email profile",
  },
  orcid: {
    // sandbox test client from orcid
    //clientId: "APP-207226CSNVZ1ZSQM",
    //authorizeUrl: "https://sandbox.orcid.org/oauth/authorize",
    
    // OAuth client: orcid "raido" account
    clientId: "APP-89ARS7MGN01HGYUO",
    authorizeUrl: "https://orcid.org/oauth/authorize",
    
    authnScope: "openid",
  },
  authApiQuery: {
    retry: 1,
    refetchOnWindowFocus: true,
  },
  publicApiQuery: {
    retry: false,
    refetchOnWindowFocus: false,
  },
};

const prodConfig: EnvironmentConfig = {
  environmentName: "prod",
  isProd: true,
  raidoIssuer: "https://api.prod.raid.org.au",
  raidoLandingPage: "https://prod.raid.org.au",
  raidoApiSvc: "",
  aaf: {
    /* This is the PROD client ID allocated by AAF via ticket
    https://aaf.freshdesk.com/helpdesk/tickets/9910 */
    clientId: "205290b0-8bd6-4d53-8605-d0237636b235",
    authorizeUrl: "https://central.aaf.edu.au/providers/op/authorize",
    authnScope: "openid email profile",
  },
  google: {
    // OAuth client: raid.service@ardc.edu.au / ardc.edu.au / Raido production / Raido PROD
    clientId: "333652356987-6n1dlouree79qbffrtgqd9fit6phkvs5.apps.googleusercontent.com",
    authorizeUrl: "https://accounts.google.com/o/oauth2/v2/auth",
    authnScope: "openid email profile",
  },
  orcid: {
    // this is the PROD client ID, created under the raid.services@ardc.eud.au ORCiD account
    clientId: "APP-89ARS7MGN01HGYUO",
    authorizeUrl: "https://orcid.org/oauth/authorize",
    authnScope: "openid",
  },
  authApiQuery: {
    retry: 1,
    refetchOnWindowFocus: true,
  },
  publicApiQuery: {
    retry: false,
    refetchOnWindowFocus: false,
  },
};


export const Config: EnvironmentConfig & typeof buildConfig = initConfig();

